### Intro to Python Programming
- Prof Joanna Bieri
- Office: Duke 209

### Day 20 - Putting the pieces together

Today we are going to work on writing some code that puts some important ideas together. This notebook will allow you to practice the following important ideas.

- Data Types
- Loops
- User Input
- Functions
- Pandas
- Plotting
- Geopandas

In [49]:
import pandas as pd
import matplotlib.pyplot as plt
import numpy as np


import geopandas as gpd
import cartopy
import cartopy.crs as ccrs

### We will start by loading some data into a Pandas Data Frame

In [50]:
website='https://joannabieri.com/python/'

file_name = website + "FruitPrices2020.csv"
DFfruit = pd.read_csv(file_name)
file_name = website + 'VegetablePrices2020.csv'
DFveg = pd.read_csv(file_name)
DF = pd.concat([DFfruit,DFveg],ignore_index=True)

### Look at the Pandas Data below

In [52]:
DF.head(5)

,Item,Form,RetailPrice,RetailPriceUnit,Yield,CupEquivalentSize,CupEquivalentUnit,CupEquivalentPrice
0,Apples,Fresh,1.5193,per pound,0.90,0.2425,pounds,0.4094
1,"Apples, applesauce",Canned,1.0660,per pound,1.00,0.5401,pounds,0.5758
2,"Apples, ready-to-drink",Juice,0.7804,per pint,1.00,8.0000,fluid ounces,0.3902
3,"Apples, frozen concentrate",Juice,0.5853,per pint,1.00,8.0000,fluid ounces,0.2926
4,Apricots,Fresh,2.9665,per pound,0.93,0.3638,pounds,1.1603


#### What are the Column Names?

In [ ]:
# Your code here


#### What are the Data Types in each Column?

You can either use the built in Pandas functions or figure it out by looking at the DataFrame.

- Check your answer using a command like:
     
      type(DF['Item'].iloc[0])
      
- What does the .iloc[0] do?

In [54]:
# Your code here


#### How many rows are in the data?

In [ ]:
# Your code here


#### Write a FOR loop that loops through each string in the Item column and prints out all the items.

In [ ]:
# Your code here


#### Write a FOR loop that loops through each string in the Item column and checks if the word "Apple" is in the string. IF "Apple" is in the string print the string.

In [ ]:
# Your code here


#### How many unique entries are there in the column "Form"?

In [ ]:
# Your code here


#### Write a FOR loop that loops through the unique entries in the column 'Form' and prints them out.

In [ ]:
# Your code here


#### Write a function that given a DataFrame will return a random row from that DataFrame.

In [ ]:
# Your code here


#### Write a function that takes as an input from a user. Ask the user for a dollar amount and then choose a random row from the data frame and prints out what the item and form are and how many pounds of that item a person could buy.

NOTE - you should be able to call the function you wrote in the step above to get the random row.



In [ ]:
# Your code here


#### Write a function that asks the user to name a fruit or vegetable and then returns all the rows in the DataFrame that contain that fruit or vegetable in their Item column (string).

NOTE - Remeber you aready wrote above to look for string matches in the items.

NOTE - My strategy for this involved creating a mask on element at a time.

NOTE - You might want to use .title() on the users input to match the format in the data.

In [ ]:
# Your code here


#### Using the function above, write a function that creates a bar plot of the DataFrame that is returned. You should plot the Item on the x-axis and the CupEquivalentPrice on the y-axis so the person can see the relative prices.

In [ ]:
# Your code here


#### CHALLENGE -  Write code that allows a user to look certain produce idems and saves their produce information list to a file that they can look at later.

- Ask the user to input a fruit or vegetable name.
- Look for that name in the strings in the DF['Item"] column
- Show the user all the rows that contain their fruit or vegetable. 
    - For example if my user gave me apple I would:
        - use .title() to make it Apple
        - search for any items that contain apple
        - create a mask so I get back just the rows about apples:
        
              0	Apples	Fresh	1.5193	per pound	0.9000	0.2425	pounds	0.4094
              1	Apples, applesauce	Canned	1.0660	per pound	1.0000	0.5401	pounds	0.5758
              2	Apples, to-drink	Juice	0.7804	per pint	1.0000	8.0000	fluid ounces  0.3902
              3	Apples, frozen conce	Juice	0.5853	per pint   1.0000	8.0000  fluid ounces 0.2926
               
- Let the user choose which row they want- eg 0,1,2,3 for apples
- Save that (.iloc[]) row to a list
- Let the user keep adding produce until they are done (WHILE loop)
- Once they are done, print the final data (DataFrame would look best) AND save the data as a .csv.

NOTE - you will want to allow for user errors like entering a word instead of an index. or doing silly things like users do.

In [80]:
# Your code here


### Look at the GeoPandas Data below

In [62]:
borderDF = pd.read_csv('Border_Crossing_Entry_Data.csv')
geometry = gpd.points_from_xy(borderDF["Longitude"], borderDF["Latitude"])
gDF = gpd.GeoDataFrame(borderDF, geometry=geometry, crs=4326)

In [63]:
gDF.head()

,Port Name,State,Port Code,Border,Date,Measure,Value,Latitude,Longitude,Point,geometry
0,Roma,Texas,2310,US-Mexico Border,Dec 2023,Buses,46,26.404,-99.019,POINT (-99.018981 26.403928),POINT (-99.01900 26.40400)
1,Del Rio,Texas,2302,US-Mexico Border,Dec 2023,Trucks,6552,29.327,-100.928,POINT (-100.927612 29.326784),POINT (-100.92800 29.32700)
2,Roma,Texas,2310,US-Mexico Border,Nov 2023,Trucks,3753,26.404,-99.019,POINT (-99.018981 26.403928),POINT (-99.01900 26.40400)
3,Douglas,Arizona,2601,US-Mexico Border,Oct 2023,Buses,13,31.334,-109.560,POINT (-109.560344 31.334043),POINT (-109.56000 31.33400)
4,Beecher Falls,Vermont,206,US-Canada Border,Aug 2023,Trucks,422,45.013,-71.505,POINT (-71.505309 45.013411),POINT (-71.50500 45.01300)


#### This data contains border crossing information for each of the ports along the US border. 

Assume that 'Value' is the total number of people crossing using that 'Measure' for that 'Date'

#### Use the GeoPandas Plotting Tools to create a plot that is similar to the image below.

In [77]:
!wget https://joannabieri.com/python/BorderFig.jpg -O BorderFig.jpg
# importing PIL
from PIL import Image
# Read image
img = Image.open('BorderFig.jpg')
# Output Images
img

In [81]:
# Your code here


#### Challenge:

See if you can ask some more interesting questions of the data and write code to get the answers.
- Which border crossing was most busy
- Which border crossing was least busy
- What form of transport brings the bost people across the border
- Create a graph of the total number of border crossings per month in the data

In [ ]:
# Your code here
